BJS2 40560 setting goals

ZhubanyshZh opened the pull request.

ZhubanyshZh committed:
feat(40560): create goal

Introduces functionality for creating user goals with title, description, deadline, and associated skills. Implements validators for user existence, active goal limits, and skill validation, along with necessary service interfaces, mappers, and controllers. Updates application configuration to include active goal limits.

ZhubanyshZh committed:
feat(40560): goal update functionality

Introduced an update endpoint for goals with validation logic to prevent updates on completed goals. Enhanced skill-user association by linking skills to users upon goal completion. Included additional constraints for goal DTOs and updated repositories, mappers, and services to support these feature.

ZhubanyshZh committed:
test: tested and fixed something

ZhubanyshZh committed:
feat(40560): delete goal by id

Added functionality to delete goals, including validation of goal ID existence and logging upon successful deletion. Introduced tests to cover both valid and invalid cases, ensuring robust behavior for goal deletion. A new endpoint was also added to `GoalController` to expose the delete functionality.

ZhubanyshZh committed:
feat(40560): add getting subtasks

Added endpoint to retrieve subtasks for a specific goal, with optional filters for title and status. Implemented service layer logic, including validation of goal existence and filtering criteria

ZhubanyshZh committed:
feat(40560): add getting user goals

added endpoint which get user goals by user ID and by filters like title and goal status and tested all of this with postman

ZhubanyshZh committed:
feat(40560): fix delete functionality

fix Goal Entity to cascade deleting

MiraGlaceon commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,63 @@
+
+    private final GoalService goalService;
+
+    @PostMapping("goal")
```
У тебя уже есть путь /goals, не делай дополнительно goal. Иначе получится /goals/goal. Оставь PostMapping пустым и тогда создание новой цели будет на пути /goals

MiraGlaceon commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,63 @@
+        return goalService.create(userId, goalDto);
+    }
+
+    @PutMapping("goal")
```
Аналогично везде подправить

MiraGlaceon commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,63 @@
+        goalService.delete(id);
+    }
+
+    @GetMapping("{id}")
```
Здесь, например, все верно

MiraGlaceon commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,63 @@
+        return goalService.findSubtasksByGoalId(id, title, status);
+    }
+
+    @GetMapping("/user/{id}")
```
Здесь нужно сохранить /goals/{id}. Ты ведь все так же ищешь цели. Если создать путь /goals/user/{id}, то я воспринимаю это как поиск человека по айдишнику. Тут нужен другой подход, либо через параметры, либо через тело запроса, посмотри как лучше

MiraGlaceon commented on src/main/java/school/faang/user_service/dto/goal/UpdateGoalDto.java:
```diff
@@ -0,0 +1,34 @@
+
+    private List<Long> skillsToAchieveIds;
+
+    @Enumerated(value = EnumType.ORDINAL)
```
Лучше не использовать ORDINAL в сущностях. DTO отражает сущность, но у тебя, представь, в базе был записан статус "2" - COMPLETED, ты в перечисление в начало добавил еще статус и у тебе вместо COMPLETED цели получилась DENIED("3"), потому что счетчик ORDINAL съехал на единицу. Это все утрировано, но когда завязываешь логику на порядке перечисления, могут быть проблемы при обновлении enum

ZhubanyshZh committed:
fix(40560): fix review checks

Changed urls of GoalController to correctly understanding and fixed UpdateGoalDto fields
